src/ILCompiler.Compiler/src/Compiler/PreInitFieldInfo.cs

```diff
@@ -8,9 +8,71 @@
 
 using Internal.TypeSystem;
 using Internal.TypeSystem.Ecma;
+using ILCompiler.DependencyAnalysis;
 
 namespace ILCompiler
 {
+    public abstract class PreInitFixupInfo : IComparable<PreInitFixupInfo>
+    {
+        /// <summary>
+        /// Offset into the blob
+        /// </summary>
+        public int Offset { get; }
+
+        public PreInitFixupInfo(int offset)
+        {
+            Offset = offset;
+        }
+
+        int IComparable<PreInitFixupInfo>.CompareTo(PreInitFixupInfo other)
+        {
+            return this.Offset - other.Offset;
+        }
+
+        /// <summary>
+        /// Writes fixup data into current ObjectDataBuilder. Caller needs to make sure ObjectDataBuilder is
+        /// at correct offset before writing.
+        /// </summary>
+        /// <returns>Bytes written</returns>
+        public abstract int WriteData(ref ObjectDataBuilder builder, NodeFactory factory);
+    }
+
+    public class PreInitTypeFixupInfo : PreInitFixupInfo
+    {
+        public TypeDesc TypeFixup { get; }
+
+        public PreInitTypeFixupInfo(int offset, TypeDesc type)
+            :base(offset)
+        {
+            TypeFixup = type;
+        }
+
+        public override int WriteData(ref ObjectDataBuilder builder, NodeFactory factory)
+        {
+            builder.EmitPointerReloc(factory.NecessaryTypeSymbol(TypeFixup));
+
+            return IntPtr.Size;
```

Comment:
Good points. I missed that we have CountBytes in ObjectDataBuilder. Removed the return and change offset code to use CountBytes.